grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -2814,4 +2816,28 @@ class StockMovementService {
         }
         return true
     }
+
+    void rollbackApproval(String stockMovementId) {
+        StockMovement stockMovement = getStockMovement(stockMovementId)
+
+        Requisition requisition = stockMovement?.requisition
+        if (requisition.status in [RequisitionStatus.APPROVED, RequisitionStatus.REJECTED]) {
+            requisitionService.rollbackLastEvent(requisition)
+            requisition.status = RequisitionStatus.PENDING_APPROVAL
+            requisition.approvedBy = null
+            requisition.rejectedBy = null
+            requisition.dateApproved = null
+            requisition.dateRejected = null
+        }
+    }
+
+    Boolean canRollbackApproval(User user, String stockMovementId) {
+        StockMovement stockMovement = getStockMovement(stockMovementId)
+
+        // When a user is an approver or requestor or is a manager or superuser, then can rollback approval
+        Boolean isApproverOrRequestor = stockMovement?.approvers?.contains(user) || stockMovement?.requestedBy == user
+        Boolean isManagerOrSuperuser = userService.isSuperuser(user) || userService.isUserAdmin(user)
```

Comment:
Manager and Admin are different roles. Do the requirements state manager or admin?

I would just interrogate the user object to see if has a role greater than manager. In fact, we could throw these methods into the domain / command classes.

Do we need to take the location into account when dealing with the user roles?
```
Boolean canUserRollbackApproval(String id, User user) {
    StockMovement stockMovement = getStockMovement(stockMovementId)     
    return stockMovement.canUserRollback(user) || userService.isUserManager()
}
```
Do we need to take the origin or destination location into account when considering roles?

One other thing we'll want to check here is whether the stock movement is in an approval state. We aren't checking whether the user is able to rollback other events, so we shouldn't allow it. Something like 

StockMovement
```
Boolean isInApprovalWorkflow() { 
    status in RequisitionStatus.listApprovalStatuses()
}
```

Reply:
> Manager and Admin are different roles. Do the requirements state manager or admin?

Uh, at the beginning, I wrote "Manager" by accident, and I forgot to change the variable name later 😕